src/Symfony/Component/Uid/Uuid.php

```diff
@@ -18,79 +18,86 @@
-    public const TYPE_4 = UUID_TYPE_RANDOM;
-    public const TYPE_5 = UUID_TYPE_SHA1;
+    protected const TYPE = UUID_TYPE_DEFAULT;
 
-    public const VARIANT_NCS = UUID_VARIANT_NCS;
-    public const VARIANT_DCE = UUID_VARIANT_DCE;
-    public const VARIANT_MICROSOFT = UUID_VARIANT_MICROSOFT;
-    public const VARIANT_OTHER = UUID_VARIANT_OTHER;
+    protected $uuid;
 
-    // https://tools.ietf.org/html/rfc4122#section-4.1.4
-    // 0x01b21dd213814000 is the number of 100-ns intervals between the
-    // UUID epoch 1582-10-15 00:00:00 and the Unix epoch 1970-01-01 00:00:00.
-    private const TIME_OFFSET_INT = 0x01b21dd213814000;
-    private const TIME_OFFSET_COM = "\xfe\x4d\xe2\x2d\xec\x7e\xc0\x00";
-
-    private $uuid;
-
-    public function __construct(string $uuid = null)
+    public function __construct(string $uuid)
     {
-        if (null === $uuid) {
-            $this->uuid = uuid_create(self::TYPE_4);
-
-            return;
+        if (static::TYPE !== uuid_type($uuid)) {
+            throw new \InvalidArgumentException(sprintf('Invalid UUID%s: "%s".', static::TYPE ? 'v'.static::TYPE : '', $uuid));
         }
 
-        if (!uuid_is_valid($uuid)) {
-            throw new \InvalidArgumentException(sprintf('Invalid UUID: "%s".', $uuid));
+        $this->uuid = strtr($uuid, 'ABCDEF', 'abcdef');
+    }
+
+    /**
+     * @return static
+     */
+    public static function fromBinary(string $bytes): self
+    {
+        if (!$uuid = uuid_unparse($bytes)) {
+            throw new \InvalidArgumentException('Invalid binary UUID provided.');
         }
 
-        $this->uuid = strtr($uuid, 'ABCDEF', 'abcdef');
+        return static::fromString($uuid);
     }
 
-    public static function v1(): self
+    /**
+     * @return static
+     */
+    public static function fromString(string $uuid): self
     {
-        return new self(uuid_create(self::TYPE_1));
+        if (__CLASS__ !== static::class) {
+            return new static($uuid);
+        }
+
+        switch (uuid_type($uuid)) {
+            case UuidV1::TYPE: return new UuidV1($uuid);
+            case UuidV3::TYPE: return new UuidV3($uuid);
+            case UuidV4::TYPE: return new UuidV4($uuid);
+            case UuidV5::TYPE: return new UuidV5($uuid);
+            case NullUuid::TYPE: return new NullUuid();
+            case self::TYPE: return new self($uuid);
+        }
+
+        throw new \InvalidArgumentException(sprintf('Invalid UUID: "%s".', $uuid));
     }
 
-    public static function v3(self $uuidNamespace, string $name): self
+    public static function v1(string $uuid = null): UuidV1
     {
-        return new self(uuid_generate_md5($uuidNamespace->uuid, $name));
+        return new UuidV1($uuid);
     }
 
-    public static function v4(): self
+    public static function v3(self $namespace, string $name): UuidV3
     {
-        return new self(uuid_create(self::TYPE_4));
+        return new UuidV3(uuid_generate_md5($namespace->uuid, $name));
```

Comment:
Have you considered moving generation logic to related Uuid classes? I.e. `UuidV3::generate($namespace, $name)`

The comment in the `UuidV3` class says: `A v3 UUID is generated by taking the MD5 of another UUID and a name.`, but the actual code that does what the comment says is somewhere else.

Reply:
We need a way to generate a V3, and a way to instantiate a V3.
Uuid::v3() is the generator, new UuidV3() the instantiator.
I updated the docblock. OK for you?